### ISSUE TITLE
Searching for nested test classes fails if class file format is invalid

### ISSUE THREAD

**Issue by tonyabbott** [junit-team/junit5#1436]
## Overview

Original issue title: _Test discovery fails when encountering invalid reference in InnerClasses attribute of class file_

The following test written in Groovy and run with JUnit 5.2.0, JUnit platform 1.2.0 results in the exception below and no tests found.

The same test, annotated with org.junit.Test and run with JUnit 4.12 successfully runs the test named foo.

I think the desirable behaviour is to log the warning as below but still continue with test discovery and run what tests are found, as JUnit 4 does.

```groovy
package tests

import org.junit.jupiter.api.Test

class GroovyInnerClassFailure {
    @Test
    void foo() {
    }

    interface Foo {
    }
}
```

Causes the following error:
```
May 29, 2018 9:46:36 AM org.junit.platform.launcher.core.DefaultLauncher handleThrowable
WARNING: TestEngine with ID 'junit-jupiter' failed to discover tests
java.lang.NoClassDefFoundError: tests/GroovyInnerClassFailure$Foo$1
	at java.lang.Class.getDeclaredClasses0(Native Method)
	at java.lang.Class.getDeclaredClasses(Class.java:1867)
	at org.junit.platform.commons.util.ReflectionUtils.findNestedClasses(ReflectionUtils.java:787)
	at org.junit.platform.commons.util.ReflectionUtils.findNestedClasses(ReflectionUtils.java:777)
	at org.junit.jupiter.engine.discovery.JavaElementsResolver.resolveContainedNestedClasses(JavaElementsResolver.java:183)
	at org.junit.jupiter.engine.discovery.JavaElementsResolver.resolveChildren(JavaElementsResolver.java:178)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.junit.jupiter.engine.discovery.JavaElementsResolver.resolveClass(JavaElementsResolver.java:61)
	at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.lambda$resolve$3(DiscoverySelectorResolver.java:69)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.resolve(DiscoverySelectorResolver.java:68)
	at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:50)
	at org.junit.jupiter.engine.JupiterTestEngine.discover(JupiterTestEngine.java:61)
	at org.junit.platform.launcher.core.DefaultLauncher.discoverEngineRoot(DefaultLauncher.java:130)
	at org.junit.platform.launcher.core.DefaultLauncher.discoverRoot(DefaultLauncher.java:117)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:82)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:52)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.ClassNotFoundException: tests.GroovyInnerClassFailure$Foo$1
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 20 more

May 29, 2018 9:46:36 AM org.junit.platform.launcher.core.DefaultLauncher handleThrowable
WARNING: TestEngine with ID 'junit-jupiter' failed to discover tests
java.lang.NoClassDefFoundError: tests/GroovyInnerClassFailure$Foo$1
	at java.lang.Class.getDeclaredClasses0(Native Method)
	at java.lang.Class.getDeclaredClasses(Class.java:1867)
	at org.junit.platform.commons.util.ReflectionUtils.findNestedClasses(ReflectionUtils.java:787)
	at org.junit.platform.commons.util.ReflectionUtils.findNestedClasses(ReflectionUtils.java:777)
	at org.junit.jupiter.engine.discovery.JavaElementsResolver.resolveContainedNestedClasses(JavaElementsResolver.java:183)
	at org.junit.jupiter.engine.discovery.JavaElementsResolver.resolveChildren(JavaElementsResolver.java:178)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.junit.jupiter.engine.discovery.JavaElementsResolver.resolveClass(JavaElementsResolver.java:61)
	at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.lambda$resolve$3(DiscoverySelectorResolver.java:69)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.resolve(DiscoverySelectorResolver.java:68)
	at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:50)
	at org.junit.jupiter.engine.JupiterTestEngine.discover(JupiterTestEngine.java:61)
	at org.junit.platform.launcher.core.DefaultLauncher.discoverEngineRoot(DefaultLauncher.java:130)
	at org.junit.platform.launcher.core.DefaultLauncher.discoverRoot(DefaultLauncher.java:117)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:74)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

The cause seems to be an erroneous reference to an undefined class in the InnerClasses attribute of the class file generated by the groovy compiler. I've logged this as https://issues.apache.org/jira/browse/GROOVY-8614

```
$ javap -v -cp out/production/classes/ tests.GroovyInnerClassFailure
...
SourceFile: "GroovyInnerClassFailure.groovy"
InnerClasses:
     public #142= #141 of #2; //Foo=class tests/GroovyInnerClassFailure$Foo of class tests/GroovyInnerClassFailure
     static #145= #144 of #2; //1=class tests/GroovyInnerClassFailure$Foo$1 of class tests/GroovyInnerClassFailure
```

## Related Issues

- #1438

**Comment by sbrannen** [comment]
Tentatively slated for 5.3 M1 for investigation

**Comment by sbrannen** [comment]
@tonyabbott, would you be willing to create minimal project that demonstrates the behavior you are witnessing?

That would be extremely helpful to the team in terms of investigating what's going on!

Either a PR for this repository or a separate project in your own public repository would suffice.

**Comment by sbrannen** [comment]
@tonyabbott, thanks for the sample project!

I can confirm that this is a bug.

We'll look into it!

**Comment by sbrannen** [comment]
FWIW, the issue still exists when using Groovy 2.5.0.

**Comment by sbrannen** [comment]
The error seems to occur in `org.junit.platform.commons.util.ReflectionUtils.findNestedClasses(Class<?>, Set<Class<?>>)`.

Specifically, we assume the invocation of `java.lang.Class.getDeclaredClasses()` will not throw an exception, but in this particular scenario that invocation indeed throws an exception. 

So we'll likely want to wrap that in a try-catch block.

**Comment by sbrannen** [comment]
Though it would likely be a better idea to handle the exceptions at a higher level such as in `org.junit.jupiter.engine.discovery.JavaElementsResolver.resolveChildren(TestDescriptor).

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
@tonyabbott, with the help of your example project and class loading magic, I've been able to reproduce the failure in a test:

```java
	@Test
	void findNestedClassesWithBrokenNestedClass() throws Exception {
		URL jarfile = getClass().getResource("/gh-1436-groovy-broken-class-file.jar");

		try (URLClassLoader classLoader = new URLClassLoader(new URL[] { jarfile })) {
			String className = "tests.NestedInterfaceGroovyTests";
			Class<?> classWithBrokenClassFile = classLoader.loadClass(className);

			assertEquals(className, classWithBrokenClassFile.getName());
			NoClassDefFoundError noClassDefFoundError = assertThrows(NoClassDefFoundError.class,
				() -> classWithBrokenClassFile.getDeclaredClasses());
			assertEquals("tests/NestedInterfaceGroovyTests$NestedInterface$1", noClassDefFoundError.getMessage());

			assertThat(ReflectionUtils.findNestedClasses(classWithBrokenClassFile, clazz -> true)).isEmpty();
		}
	}
```

So that'll make it easier to fix (and know it's fixed). 😉

**Comment by sbrannen** [comment]
Current work on this issue can be viewed in the following branch:

https://github.com/junit-team/junit5/tree/issues/1436-nested-class-with-invalid-class-file

**Comment by sbrannen** [comment]
FYI: I've limited the scope of _this_ GitHub issue to address the lack of error handling in `ReflectionUtils` with regard to class loading for _nested classes_, and I have opened #1438 in order to address the fact that the Jupiter test engine aborts the discovery process when certain kinds of exceptions are encountered.

**Comment by sbrannen** [comment]
Fixed in d0e31388252275cc873d8d0bcb907331cec89f67 and d0b759229b5a0c4eb62d95a75fe61f980482b38f
